### ISSUE TITLE
Can preconstructed indices be trained for a reference dataset and then subsetting for common genes before testing? 

### ISSUE THREAD

**Issue by mauritsunkel** [LTLA/SingleR#222]
Hi again, thanks for pointing me to the book last time: http://bioconductor.org/books/release/SingleRBook/advanced-options.html#advanced-options, very informative for splitting the training and testing phases. Here it is explicitly stated that "The resulting data structure can then be re-used across multiple classifications with different test datasets, provided the gene annotation in the test dataset is identical to or a superset of the genes in the training set."
1) Why is this? 

However, as I have multiple test datasets (with different common gene spaces), I would want to train on our reference dataset only once to preconstruct indices.

2) Is it still possible at all to perform gene intersection after training, so before testing? 
3) If it is possible, does it make any sense to do it in this order, and if not, why not? 

Thanks in advance.

**Comment by LTLA** [comment]
> 1. Why is this?

The current implementation identifies the top X markers from the reference as part of the training process. That assumes that all of those markers are available in the test dataset, hence the need of an identical set or a superset.

It's not easy to perform _post-hoc_ subsetting to common genes, as the training process immediately uses the top X markers to construct ranked expression profiles and to build neighbor search indices. This would need to be redone if any of the X markers was not available in the test dataset, which would defeat the purpose of a separate training step.

> 2. Is it still possible at all to perform gene intersection after training, so before testing?

No. The construction of ranked expression profiles would change if any genes were lost, so that would need to be repeated, and so would the index building.

The best that can be done is to splitt off the marker detection into a separate step, where each pairwise comparison between labels yields a ranking of marker genes rather than just the top set. Then, the application of **SingleR** can just use the top X markers _after_ taking the intersection with the test dataset. This means that the marker detection is only done once. But there's no getting around the fact that the rest of the training needs to be repeated per test.

**Comment by mauritsunkel** [comment]
Very clear answer, thank you!
